File: src/lib/review-window.ts
/**
 * Review Today only judges blocks that have actually had their chance.
 *
 * A pending block whose end is still ahead of "now" has not failed to happen
 * yet, so it must not be offered to "let go" or "move to tomorrow". Those are
 * counted (so the screen can say they are still coming) but never listed.
 *
 * `nowMin` is minutes from midnight in the planning zone; `null` means the day
 * under review is not today, in which case every pending block is in scope.
 */
export function partitionReviewItems<
  T extends { startMin: number; durationMin: number },
>(items: T[], nowMin: number | null): { past: T[]; upcoming: number } {
  if (nowMin == null) return { past: items, upcoming: 0 };
  const past = items.filter((item) => item.startMin + item.durationMin <= nowMin);
  return { past, upcoming: items.length - past.length };
}

File: ios/App/Features/Today/ReviewWindow.swift
import Foundation

/// Review Today only judges blocks that have actually had their chance.
/// Mirrors `src/lib/review-window.ts` so web and native cannot diverge.
enum ReviewWindow {
    static func partition<T>(
        _ items: [T],
        nowMin: Int?,
        startMin: (T) -> Int,
        durationMin: (T) -> Int
    ) -> (past: [T], upcoming: Int) {
        guard let nowMin else { return (items, 0) }
        let past = items.filter { startMin($0) + durationMin($0) <= nowMin }
        return (past, items.count - past.count)
    }

    static func partition(_ items: [DayBlock], nowMin: Int?) -> (past: [DayBlock], upcoming: Int) {
        partition(
            items,
            nowMin: nowMin,
            startMin: { $0.startMin },
            durationMin: { $0.durationMin }
        )
    }
}
